freebsd/utils/jail.cpp

```diff
@@ -0,0 +1,323 @@
+                quote = 0;
+            }
+            else
+                p += c;
+        }
+    }
+
+    // leftovers
+    if (!p.empty())
+        params.push_back(p);
+
+    return params;
+}
+
+
+/// Constructs a jail name based on program and test case.
+///
+/// The formula is "kyua" + <program path> + "_" + <test case name>.
+/// All non-alphanumeric chars are replaced with "_".
+///
+/// If a resulting string exceeds maximum allowed length of a jail name,
+/// then it's shortened from the left side keeping the "kyua" prefix.
+///
+/// \param program The test program.
+/// \param test_case_name Name of the test case.
+///
+/// \return A jail name string.
+std::string
+jail::make_name(const fs::path& program,
+                const std::string& test_case_name)
+{
+    std::string name = std::regex_replace(
+        program.str() + "_" + test_case_name,
+        std::regex(R"([^A-Za-z0-9_])"),
+        "_");
+
+    const std::string::size_type limit =
+        jail_name_max_len - strlen(jail_name_prefix);
+    if (name.length() > limit)
+        name.erase(0, name.length() - limit);
+
+    return jail_name_prefix + name;
+}
+
+
+/// Create a jail with a given name and params string.
+///
+/// A new jail will always be 'persist', thus the caller is expected to remove
+/// the jail eventually via remove().
+///
+/// It's expected to be run in a subprocess.
+///
+/// \param jail_name Name of a new jail.
+/// \param jail_params String of jail parameters.
+void
+jail::create(const std::string& jail_name,
+             const std::string& jail_params)
+{
+    args_vector av;
+
+    // creation flag
+    av.push_back("-qc");
+
+    // jail name
+    av.push_back("name=" + jail_name);
+
+    // determine maximum allowed children.max
+    int max;
+    size_t len = sizeof(max);
+    if (::sysctlbyname("security.jail.children.max", &max, &len, NULL, 0) != 0) {
+        std::cerr << "sysctlbyname(security.jail.children.max) errors: "
+            << strerror(errno) << ".\n";
+        std::exit(EXIT_FAILURE);
+    }
+    if (len < sizeof(max)) {
+        std::cerr << "sysctlbyname(security.jail.children.max) provides less "
+            "data (" << len << ") than expected (" << sizeof(max) << ").\n";
+        std::exit(EXIT_FAILURE);
+    }
+    if (max < 0) {
+        std::cerr << "sysctlbyname(security.jail.children.max) yields "
+            "abnormal " << max << ".\n";
+        std::exit(EXIT_FAILURE);
+    }
+    if (max > 0)
+        max--; // a child jail must have less than parent's children.max
+    av.push_back("children.max=" + std::to_string(max));
+
+    // test defined jail params
+    const std::vector< std::string > params = parse_params_string(jail_params);
+    for (const std::string& p : params)
+        av.push_back(p);
+
+    // it must be persist
+    av.push_back("persist");
+
+    // invoke jail
+    std::auto_ptr< process::child > child = child::fork_capture(
+        run(fs::path("/usr/sbin/jail"), av));
```

Comment:
It seems like sanity checking that the jail(8) program exists and skipping with a helpful message would be wise--instead of just marking all of the tests broken.
This is especially important for systems where `WITHOUT_JAIL=` is specified in src.conf(5).

Reply:
I think that would be ok.